File: week1/ps1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long get_cc(void);
int dig_sum(int);
int checksum(long num);
int index(long, int);

int main(void)
{
    // Request cc number
    long num = get_cc();

    // Check the type
    checksum(num);
}

long get_cc(void)
{
    // request cc number until a number input is received
    long num;
    do
    {
        num = get_long("Number: ");
    }
    while (num < 1);

    return (num);
}


int checksum(long num)
{
    // Implement Luhn's algorithm
    int i = 0;
    int digit;
    long res = num;
    long sum = 0;

    while (res > 0)
    {
        i++;              // digit index
        digit = res % 10; // extract digit
        res = res / 10;   // move to next digit

        // Odd index digits are added (odd numbers have remainder)
        // Even index digits are multipled by 2 and added
        if (i % 2 == 1)
        {
            sum = sum + digit;
        }
        else
        {
            sum = sum + dig_sum(2 * digit);
        }
    }

    // Get second digit to check for card type
    int sec = index(num, i - 1);

    // Check that the number is valid, and if so
    // determine what type of card it is
    if (sum % 10 == 0)
    {
        if (((digit == 3) & (sec == 4 || sec == 7)) & (i == 15))
        {
            printf("AMEX\n");
        }
        else if ((digit == 4) & (i == 13 || i == 16))
        {
            printf("VISA\n");
        }
        else if (((digit == 5) & (0 < sec & sec < 6)) & (i == 16))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return (sum);
}

// Function to add together digits of a number
// (max value of num in this exercise is 2*19)
int dig_sum(int num)
{
    int sum = 0;
    while (num > 0)
    {
        sum = sum + num % 10;
        num = num / 10;
    }
    return (sum);
}

// Negative index function
// Counting backwards, get value in ith slot
int index(long num, int loc)
{
    int val;

    for (int i  = 0; i < loc; i++)
    {
        val = num % 10; // extract digit
        num = num / 10;   // move to next digit
    }
    return (val);
}

File: week6/sentimental-credit/credit.py
import sys


def main():
    # Request cc number - easiest to get it as a string
    num = input("Number: ")

    # length of cc number is 13, 15, 16
    # we can assume all numeric
    # need to double check, this might not be how implementation is desired
    while len(num) < 1:
        num = input("Number: ")
    # Check the number
    checksum(num)


def dig_sum(num):
    """Function to add together digits of a number"""
    sum = 0
    num = str(num)
    for i in range(len(num)):
        sum += int(num[i])
    return sum


def checksum(num):
    # Implement Luhn's algorithm
    # 1) Multiply every other digit by 2, starting with the number’s
    #    second-to-last digit, and then add those products’ digits together.
    # 2) Add the sum to the sum of the digits that weren’t multiplied by 2.
    # 3) If the total’s last digit is 0 (or, put more formally, if the total
    #    modulo 10 is congruent to 0), the number is valid!
    sum = 0
    rev = num[::-1]  # reversed string
    for i in range(len(rev)):
        if i % 2 == 0:
            sum += int(rev[i])
        else:
            sum += dig_sum(2 * int(rev[i]))

    # Check type
    if sum % 10 != 0:
        print("INVALID")
    elif len(num) == 15 and num[0] == "3" and (num[1] == "4" or num[1] == "7"):
        # Amex has 15 digits and starts with 34 or 37
        print("AMEX")
    elif (len(num) == 16 or len(num) == 13) and num[0] == "4":
        # Visa has 13 or 16 length and starts with 4
        print("VISA")
    elif len(num) == 16 and int(num[0:2]) in range(51, 56):
        # MasterCard has 16 digits and start with 51, 52, 53, 54, or 55
        print("MASTERCARD")
    else:
        print("INVALID")

    return sum


if __name__ == "__main__":
    main()
